chain/network/src/peer/peer_actor.rs

```diff
@@ -1117,22 +1117,29 @@ impl PeerActor {
                     .network_state
                     .peer_store
                     .healthy_peers(self.network_state.config.max_send_peers as usize);
-                if !peers.is_empty() {
-                    tracing::debug!(target: "network", "Peers request from {}: sending {} peers.", self.peer_info, peers.len());
-                    self.send_message_or_log(&PeerMessage::PeersResponse(peers));
+                let direct_peers = self.network_state.get_direct_peers();
+                if !peers.is_empty() || !direct_peers.is_empty() {
+                    tracing::debug!(target: "network", "Peers request from {}: sending {} peers and {} direct peers.", self.peer_info, peers.len(), direct_peers.len());
+                    self.send_message_or_log(&PeerMessage::PeersResponse(peers, direct_peers));
                 }
                 self.network_state
                     .config
                     .event_sink
                     .push(Event::MessageProcessed(conn.tier, peer_msg));
             }
-            PeerMessage::PeersResponse(peers) => {
-                tracing::debug!(target: "network", "Received peers from {}: {} peers.", self.peer_info, peers.len());
+            PeerMessage::PeersResponse(peers, direct_peers) => {
+                tracing::debug!(target: "network", "Received peers from {}: {} peers and {} direct peers.", self.peer_info, peers.len(), direct_peers.len());
                 let node_id = self.network_state.config.node_id();
                 self.network_state.peer_store.add_indirect_peers(
                     &self.clock,
                     peers.into_iter().filter(|peer_info| peer_info.id != node_id),
                 );
+                // Direct peers of the responding peer are still indirect peers for this node.
+                // However, we may treat them with more trust in the future.
+                self.network_state.peer_store.add_indirect_peers(
+                    &self.clock,
+                    direct_peers.into_iter().filter(|peer_info| peer_info.id != node_id),
```

Comment:
while you are at it, perhaps move this filtering condition inside peer_store?

Reply:
As discussed offline we will leave this as it is because peer_store does not know the node_id.

Reply:
As discussed offline, peer_store does not know the node_id. I agree that having the filtering here is a little unfortunate but I am leaving it as is for now.